ogre2/src/Ogre2BoundingBoxCamera.cc

```diff
@@ -0,0 +1,1078 @@
+      uint32_t ogreId1 = this->dataPtr->buffer[index + 1];
+      uint32_t ogreId2 = this->dataPtr->buffer[index + 0];
+
+      if (label != this->dataPtr->materialSwitcher->backgroundLabel)
+      {
+        // get the OGRE id of 16 bit value
+        uint32_t ogreId = ogreId1 * 256 + ogreId2;
+
+        BoundingBox *box;
+        BoxBoundary *boundary;
+
+        // create new boxes when its first pixel appears
+        if (!this->dataPtr->boundingboxes.count(ogreId))
+        {
+          box = new BoundingBox(BoundingBoxType::VisibleBox2D);
+          box->label = label;
+
+          boundary = new BoxBoundary();
+          boundary->minX = width;
+          boundary->minY = height;
+          boundary->maxX = 0;
+          boundary->maxY = 0;
+
+          boxesBoundary[ogreId] = boundary;
+          this->dataPtr->boundingboxes[ogreId] = box;
+        }
+        else
+        {
+          boundary = boxesBoundary[ogreId];
+        }
+
+        boundary->minX = std::min<uint32_t>(boundary->minX, x);
+        boundary->minY = std::min<uint32_t>(boundary->minY, y);
+        boundary->maxX = std::max<uint32_t>(boundary->maxX, x);
+        boundary->maxY = std::max<uint32_t>(boundary->maxY, y);
+      }
+    }
+  }
+
+  for (auto box : this->dataPtr->boundingboxes)
+  {
+    // Get the box's boundary
+    auto ogreId = box.first;
+    auto boundary = boxesBoundary[ogreId];
+    auto boxWidth = boundary->maxX - boundary->minX;
+    auto boxHeight = boundary->maxY - boundary->minY;
+
+    box.second->center.X() = boundary->minX + boxWidth / 2;
+    box.second->center.Y() = boundary->minY + boxHeight / 2;
+    box.second->center.Z() = 0;
+
+    box.second->size.X() = boxWidth;
+    box.second->size.Y() = boxHeight;
+    box.second->size.Z() = 0;
+  }
+
+  // Combine boxes of multi-links model if exists
+  this->MergeMultiLinksModels();
+}
+
+/////////////////////////////////////////////////
+void Ogre2BoundingBoxCamera::FullBoundingBoxes()
+{
+  // used to filter the hidden boxes
+  this->MarkVisibleBoxes();
+
+  Ogre::Matrix4 viewMatrix = this->ogreCamera->getViewMatrix();
+  Ogre::Matrix4 projMatrix = this->ogreCamera->getProjectionMatrix();
+
+  auto itor = this->scene->OgreSceneManager()->getMovableObjectIterator(
+      Ogre::ItemFactory::FACTORY_TYPE_NAME);
+  while (itor.hasMoreElements())
+  {
+    Ogre::MovableObject *object = itor.peekNext();
+    Ogre::Item *item = static_cast<Ogre::Item *>(object);
+    Ogre::MeshPtr mesh = item->getMesh();
+
+    uint32_t ogreId = item->getId();
+
+    // Skip the items which is hidden in the ogreId map
+    if (!this->dataPtr->visibleBoxesLabel.count(ogreId))
+    {
+      itor.moveNext();
+      continue;
+    }
+
+    // get attached node
+    Ogre::Node *node = item->getParentNode();
+
+    // get the derived position
+    Ogre::Vector3 position = node->_getDerivedPositionUpdated();
+    Ogre::Quaternion oreintation = node->_getDerivedOrientationUpdated();
```

Comment:
is there any problem with using the internal functions of OGRE ??

Reply:
I'm not sure, @iche033 what are your thoughts? I would think that we should try to shy away from internal OGRE methods (i.e., ones that start with `_`), but it does seem like they're being used in a few other places in `ign-rendering`.

Reply:
these should be fine. If possible, use the `_getDerivedPosition` (without `Updated`) version to avoid extra node updates.